### ISSUE TITLE
Implicit configuration and cabal files

### ISSUE THREAD

**Issue by jrp2014** [mpickering/hie-bios#174]
A common use case is having a vanilla cabal project and getting hie-bios to pick up right ghc flags.

It doesn't work for me, presumably because my project doesn't compile yet, so cabal v2-repl doesn't give hie-bios what it needs.  It is fixable by adding an hie.yaml file, saying please use cabal, but that is clunky and I'm not sure why it tells hie-bios anything more than it could already work out for itself.

**Comment by fendor** [comment]
The implicit config is essentially looking for a *.cabal file and then using `cabal repl File.hs` to find the options. This is pretty much equivalent to a `hie.yaml` such as:
```yaml
cradle:
  cabal:
```

What did you write in your `hie.yaml`?
Did it discover a different cradle or did it just fail to load?

**Comment by jrp2014** [comment]
Sure.  It works if I just use a hie.yaml cabal file like that in the readme here, but it doesn't seem to use the cabal fille if the hie.yaml isn't there.  I thought that the point was that it should but only if cabal v2-repl works, which it may not if the project doesn't compile.  But I'm speculating.

**Comment by fendor** [comment]
Please add some more details, what was the project you were trying to use? Did you try, e.g. `hie-bios debug Lib.hs` to see the details what cradle is selected, etc...
Why do you think it doesnt use cabal?

**Comment by Avi-D-coder** [comment]
@jrp2014 what happens if you use [implicit-hie](https://github.com/Avi-D-coder/implicit-hie) generated `hie.yaml`?

**Comment by jrp2014** [comment]
With just a '.cabal' file and the following 'hie.yaml':

```
cradle: {cabal: {component: "lib:smuggler"}}
```
I get, for `hie-bios debug`
```
Root directory:      /Users/.../Haskell/smuggler
GHC options:         -fbuilding-cabal-package -O0 -outputdir ...
Config Location:     No explicit config found
Cradle:              Cradle {cradleRootDir = "/Users/.../Haskell", cradleOptsProg = CradleAction: Default}
Dependencies:        smuggler.cabal cabal.project
 ```
That output also contains some random characters, it seems, which makes it harder to cut & paste into here, but that's a different issue.

Without that `hie-bios` file I get
```
Root directory:      /Users/.../Haskell/smuggler
GHC options:         
System libraries:    /usr/local/Cellar/ghc/8.8.3/lib/ghc-8.8.3
Config Location:     No explicit config found
Cradle:              Cradle {cradleRootDir = "/Users/.../Haskell", cradleOptsProg = CradleAction: Default}
Dependencies:        

```

**Comment by Avi-D-coder** [comment]
Please post your cabal file

**Comment by fendor** [comment]
@jrp2014 You should pass a filepath to `hie-bios debug` e.g. `src/Lib.hs`.

**Comment by jrp2014** [comment]
(I should say that my expectation was that the two outputs should be the same.)

Here's the cabal file.  (NB: It's not the same as the original smuggler one pointed to as the source repo' its a fork.)

```
cabal-version:   2.4
name:            smuggler
version:         0.2.1.0
synopsis:        GHC Source Plugin that helps to manage imports
description:
  == Usage
  .
  Add @smuggler@ to the dependencies of your project.
  .
  Then add the following to ghc-options: @-fplugin=Smuggler.Plugin@

homepage:        https://github.com/kowainik/smuggler
bug-reports:     https://github.com/kowainik/smuggler/issues
license:         MPL-2.0
license-file:    LICENSE
author:          Dmitrii Kovanikov, Veronika Romashkina
maintainer:      Kowainik <xrom.xkov@gmail.com>
copyright:       2018-2019 Kowainik
category:        Development, Refactoring
build-type:      Simple
extra-doc-files:
  README.md
  CHANGELOG.md

tested-with:     GHC ==8.8.3

flag debug
  description: Enable debugging support
  default:     False
  manual:      True

source-repository head
  type:     git
  location: git@github.com:kowainik/smuggler.git

common common-options
  build-depends:      base ^>=4.13
  ghc-options:
    -Wall -Wextra -Wincomplete-uni-patterns -Wincomplete-record-updates
    -Wcompat -Widentities -Wredundant-constraints -fhide-source-paths

  if flag(debug)
    ghc-options: -ddump-minimal-imports

  default-language:   Haskell2010
  default-extensions:
    DeriveGeneric
    GeneralizedNewtypeDeriving
    InstanceSigs
    LambdaCase
    OverloadedStrings
    RecordWildCards
    ScopedTypeVariables
    StandaloneDeriving
    TypeApplications

library
  import:          common-options
  hs-source-dirs:  src
  exposed-modules:
    Smuggler.Anns
    Smuggler.Export
    Smuggler.Import
    Smuggler.Loc
    Smuggler.Options
    Smuggler.Parser
    Smuggler.Plugin

  if flag(debug)
    exposed-modules:
      Smuggler.Debug
      Smuggler.Name

  -- Does not compile with 8.8.3      Smuggler.Import
  build-depends:
    , containers      >=0.5   && <0.7
    , filepath        ^>=1.4
    , ghc             ^>=8.8.0
    , ghc-exactprint  ^>=0.6
    , syb

  if flag(debug)
    build-depends:
      , fmt
      , pretty-simple
      , syb
      , text

executable play-smuggler
  import:         common-options
  hs-source-dirs: app
  main-is:        Main.hs
  ghc-options:    -threaded -rtsopts -with-rtsopts=-N

  if flag(debug)
    ghc-options: -fplugin=Smuggler.Plugin

  build-depends:  smuggler

common common-tests
  import:         common-options
  hs-source-dirs: test
  other-modules:
    --Test.BoolBoolUnused
    --Test.BoolBoolUsed
    --Test.BoolPartlyUsedFirst
    --Test.BoolPartlyUsedLast
    --Test.ConstructorsUnused
    --Test.Dummy
    --Test.ExportExplicit
    --Test.ExportImplicit
    Test.ImplicitImportUnused
    --Test.ImplicitImportUsed
    --Test.ImportEmpty
    --Test.MultilineUnused
    --Test.TypeConstructorUnused
    --Test.TypeConstructorUnusedComma
    --Test.TypeConstructorUsed
    --Test.TypeUnused
    --Test.TypeUsed
    --Test.WildcardUnused

  build-depends:
    , ansi-terminal
    , directory
    , filepath
    , smuggler

  ghc-options:
    -threaded -rtsopts -with-rtsopts=-N -fno-warn-unused-imports
    -fno-warn-missing-signatures -fplugin=Smuggler.Plugin
    -fplugin-opt=Smuggler.Plugin:test

test-suite smuggler-test
  import:  common-tests
  type:    exitcode-stdio-1.0
  main-is: Test.hs
  ghc-options:
    -fplugin-opt=Smuggler.Plugin:test
 --   -fplugin-opt=Smuggler.Plugin:NoExportProcessing
 --    -fplugin-opt=Smuggler.Plugin:MinimiseImports

test-suite smuggler-test-maximal
  import:  common-tests
  type:    exitcode-stdio-1.0
  main-is: Test.hs
  ghc-options:
    -fplugin-opt=Smuggler.Plugin:test
    -fplugin-opt=Smuggler.Plugin:MinimiseImports
--    -fplugin-opt=Smuggler.Plugin:ReplaceExports
```

**Comment by jrp2014** [comment]
> @jrp2014 You should pass a filepath to `hie-bios debug` e.g. `src/Lib.hs`.

That seems to produce identical output without the hie.yaml file.

**Comment by fendor** [comment]
@jrp2014 Yeah it should and the CradleAction should be `Cabal` in both cases.

What are you trying to do and what fails exactly?

**Comment by Avi-D-coder** [comment]
What happens if you use this hie.yaml
```yaml
cradle:
  cabal:
    - path: "src"
      component: "lib:smuggler"
    - path: "play-smuggler"
      component: "exe:app"
    - path: "smuggler-test"
      component: "test:."
    - path: "smuggler-test-maximal"
      component: "test:."

```

**Comment by jrp2014** [comment]
> What are you trying to do and what fails exactly?

What I am trying to do is fix this  https://github.com/dense-analysis/ale/blob/master/ale_linters/haskell/cabal_ghc.vim linter for `ale`, a `vim` plugin.

It seems to be broken because it doesn't supply a project root directory and even when you fix it to do that, ghc complains because it cannot find the hidden package `smuggler`, presumably because is trying to compile a temporary version version of the file you are editing (outside the project).  It is easy to fix for a specific project: just add `-package smuggler` for example, to the ghc command that it uses and it seems to work well.

I was hoping that `hie-bios` would generate the right information (project root, ghc flags) to make the linter go.  If I have to create a seemingly redundant `hie.yaml` file to get things going, I may as well just hack `cabal_ghc.vim` for each project and be done.

@Avi-D-coder The minimal `hie.yaml` file seems to work OK.  My object is not to have to create it, when `hie-bios` should just use the cabal file that the minimal `hie.yaml` file already points to.

**Comment by fendor** [comment]
Fixed by #221
